### PR TITLE
Fix #311: Improve detection of failed presence check

### LINKED ISSUE #311: Improve detection of failed presence check
The following logic should be improved:

```
 if (StringUtils.isNotBlank(errorDetail) || StringUtils.isNotBlank(rejectReason)) {
```

We should store the origin of the error to differentiate that the error was caused e.g. by presence check.

### PR DESCRIPTION
**Pull request by romanstrobl**
This pull requests adds an `error_origin` column which is used for auditing purposes as well as for checking overall OTP verification result during SCA step (failure due to presence check error).

DDL update:

```
alter table es_document_result add column error_origin VARCHAR(256);
alter table es_document_verification add column error_origin VARCHAR(256);
alter table es_identity_verification add column error_origin VARCHAR(256);
alter table es_onboarding_otp add column error_origin VARCHAR(256);
alter table es_onboarding_process add column error_origin VARCHAR(256);

alter table es_document_result add column reject_origin VARCHAR(256);
alter table es_document_verification add column reject_origin VARCHAR(256);
alter table es_identity_verification add column reject_origin VARCHAR(256);
```

### CODE REVIEW

File: enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/IdentityVerificationStatusService.java [hunk 1]
```diff
@@ -255,6 +259,7 @@ private void evaluatePresenceCheckResult(OwnerId ownerId,
                 break;
             case REJECTED:
                 idVerification.setRejectReason(result.getRejectReason());
+                idVerification.setErrorOrigin(ErrorOrigin.PRESENCE_CHECK);
```
Comment (saalistaja) [inline enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/IdentityVerificationStatusService.java:262]: Maybe we should consider introducing `rejectOrigin` to not mix failed error and rejected reason together on the `errorOrigin` field.

File: enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/document/DocumentProcessingService.java [hunk 2]
```diff
@@ -200,6 +202,7 @@ public void checkDocumentSubmitWithProvider(OwnerId ownerId, DocumentResultEntit
         }
 
         documentResultEntity.setErrorDetail(docSubmitResult.getErrorDetail());
```
Comment (saalistaja) [inline enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/document/DocumentProcessingService.java:204]: There seems to be a legacy mixture of error and reject reason (an old relic).

Let's consider something like
```
if (docSubmitResult.getErrorDetail() != null) {
  documentResultEntity.setErrorDetail(docSubmitResult.getErrorDetail());
  documentResultEntity.setErrorOrigin(ErrorOrigin.DOCUMENT_VERIFICATION);
}
```

File: enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/document/DocumentProcessingService.java [hunk 3]
```diff
@@ -369,6 +373,7 @@ private void processDocsSubmitResults(OwnerId ownerId, DocumentVerificationEntit
         if (docSubmitResult.getErrorDetail() != null) {
             docVerification.setStatus(DocumentStatus.FAILED);
             docVerification.setErrorDetail(docSubmitResult.getErrorDetail());
+            docVerification.setErrorOrigin(ErrorOrigin.DOCUMENT_VERIFICATION);
         } else if (docSubmitResult.getRejectReason() != null) {
             docVerification.setStatus(DocumentStatus.REJECTED);
             docVerification.setRejectReason(docSubmitResult.getRejectReason());
```
Comment (saalistaja) [inline enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/document/DocumentProcessingService.java:379]: Here we should obviously set also the `errorOrigin`?

Basically if there will be only the `errorOrigin` we should use it at all places where we set the `rejectReason` too.

File: enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/verification/VerificationProcessingService.java [hunk 4]
```diff
@@ -188,6 +187,7 @@ private void updateDocVerification(OwnerId ownerId, DocumentVerificationEntity d
     private void updateDocumentResult(DocumentResultEntity docResult,
                                       DocumentVerificationResult docVerificationResult) {
         docResult.setErrorDetail(docVerificationResult.getErrorDetail());
```
Comment (saalistaja) [inline enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/verification/VerificationProcessingService.java:189]: An old relict with mixing error and reject reason.

```
if (docVerificationResult.getErrorDetail() != null) {
  docResult.setErrorDetail(docVerificationResult.getErrorDetail());
  docResult.setErrorOrigin(ErrorOrigin.DOCUMENT_VERIFICATION);
}
if (docVerificationResult. getRejectReason() != null) {
  docResult. setRejectReason(docVerificationResult. getRejectReason());
  docResult.setErrorOrigin(ErrorOrigin.DOCUMENT_VERIFICATION);
}
```

**Review by saalistaja — CHANGES_REQUESTED** [review]
Nice addition of a new field to track error/rejection origin.

I am just not sure with the shared naming, used for `rejectReason` too.

Furthermore rhere seems to be left several places in the code with `setRejectReason` but without setting an origin, is this desired?

**Comment by romanstrobl** [comment]
Ok, we can extend the logic for the sake of completeness:
- add `reject_origin` column
- always set `reject_origin` when setting `reject_reason`
- always separate handling for rejection and errors

**Comment by romanstrobl** [comment]
@saalistaja Please check the changes now. We will also need to merge the changes with state machine logic, I can try to do it, but I think it will be easier for you?

File: docs/sql/oracle/create-schema.sql [hunk 5]
```diff
@@ -77,7 +79,9 @@ CREATE TABLE ES_IDENTITY_VERIFICATION (
     STATUS VARCHAR2(32 CHAR) NOT NULL,
     PHASE VARCHAR2(32 CHAR) NOT NULL,
     REJECT_REASON CLOB,
+    REJECT_ORIGIN VARCHAR2(256),
```
Comment (saalistaja) [inline docs/sql/oracle/create-schema.sql:82]: missing ` CHAR`?

File: docs/sql/oracle/create-schema.sql [hunk 6]
```diff
@@ -105,7 +109,9 @@ CREATE TABLE ES_DOCUMENT_VERIFICATION (
     PHOTO_ID VARCHAR2(256 CHAR),
     VERIFICATION_SCORE INTEGER,
     REJECT_REASON CLOB,
+    REJECT_ORIGIN VARCHAR2(256),
```
Comment (saalistaja) [inline docs/sql/oracle/create-schema.sql:112]: same here, `256 CHAR`?

File: docs/sql/oracle/create-schema.sql [hunk 7]
```diff
@@ -141,8 +147,10 @@ CREATE TABLE ES_DOCUMENT_RESULT (
     DOCUMENT_VERIFICATION_ID VARCHAR2(36 CHAR) NOT NULL,
     PHASE VARCHAR2(32 CHAR) NOT NULL,
     REJECT_REASON CLOB,
+    REJECT_ORIGIN VARCHAR2(256),
```
Comment (saalistaja) [inline docs/sql/oracle/create-schema.sql:150]: dtto

File: enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/IdentityVerificationOtpService.java [hunk 8]
```diff
@@ -175,9 +175,10 @@ private OtpVerifyResponse verifyPresenceCheck(final OnboardingProcessEntity proc
 
         final IdentityVerificationEntity idVerification = getIdentityVerificationEntity(process);
         final String errorDetail = idVerification.getErrorDetail();
+        final ErrorOrigin errorOrigin = idVerification.getErrorOrigin();
         final String rejectReason = idVerification.getRejectReason();
 
-        if (StringUtils.isNotBlank(errorDetail) || StringUtils.isNotBlank(rejectReason)) {
+        if (errorOrigin == ErrorOrigin.PRESENCE_CHECK && (StringUtils.isNotBlank(errorDetail) || StringUtils.isNotBlank(rejectReason))) {
```
Comment (saalistaja) [inline enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/IdentityVerificationOtpService.java:181]: Shouldn't we fix this to
`(errorOrigin == ErrorOrigin.PRESENCE_CHECK && StringUtils.isNotBlank(errorDetail)) || (rejectOrigin == ErrorOrigin.PRESENCE_CHECK && StringUtils.isNotBlank(rejectReason))` ?

File: enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/document/DocumentProcessingService.java [hunk 9]
```diff
@@ -199,9 +198,16 @@ public void checkDocumentSubmitWithProvider(OwnerId ownerId, DocumentResultEntit
             docSubmitResult.setErrorDetail(e.getMessage());
         }
 
-        documentResultEntity.setErrorDetail(docSubmitResult.getErrorDetail());
+        if (docSubmitResult.getErrorDetail() != null) {
```
Comment (saalistaja) [inline enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/document/DocumentProcessingService.java:201]: Maybe we should use `StringUtils.isNotBlank` to be consistent with other occurrences.

**Review by saalistaja — CHANGES_REQUESTED** [review]
Last round of review. I am not sure with one condition.

File: enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/IdentityVerificationOtpService.java (hunk 8 above)
Comment (romanstrobl) [reply enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/IdentityVerificationOtpService.java:181]: Yes, true.

**Review by saalistaja — APPROVED** [review]
Well done!

**Pull request merged**